analysis/preprocess/preprocess_data.R

```diff
@@ -0,0 +1,157 @@
+# Load libraries ---------------------------------------------------------------
+
+library(magrittr)
+library(dplyr)
+library(tidyverse)
+library(lubridate)
+library(data.table)
+library(readr)
+
+# Specify command arguments ----------------------------------------------------
+
+args <- commandArgs(trailingOnly=TRUE)
+print(length(args))
+if(length(args)==0){
+  cohort_name <- "vax"
+} else {
+  cohort_name <- args[[1]]
+}
+
+# Get column names -------------------------------------------------------------
+
+all_cols <- fread(paste0("output/input_",cohort_name,".csv.gz"), 
+                  header = TRUE, sep = ",", nrows = 0, 
+                  stringsAsFactors = FALSE) %>%
+  names()
+
+message("Column names found")
+
+# Identify column classes ------------------------------------------------------
+
+cat_cols <- c("patient_id", grep("_cat", all_cols, value = TRUE))
+
+bin_cols <- c(grep("_bin", all_cols, value = TRUE),
+              "has_follow_up_previous_6months", "was_alive", "has_died")
+
+num_cols <- c(grep("_num", all_cols, value = TRUE),
+              grep("vax_jcvi_age_", all_cols, value = TRUE))
+
+date_cols <- grep("_date", all_cols, value = TRUE)
+
+message("Column classes identified")
+
+# Define column classes --------------------------------------------------------
+
+col_classes <- setNames(
+  c(rep("c", length(cat_cols)),
+    rep("l", length(bin_cols)),
+    rep("d", length(num_cols)),
+    rep("D", length(date_cols))
+  ), 
+  all_cols[match(c(cat_cols, bin_cols, num_cols, date_cols), all_cols)]
+)
+
+message("Column classes defined")
+
+# Read cohort dataset ---------------------------------------------------------- 
+
+df <- read_csv(paste0("output/input_",cohort_name,".csv.gz"), 
+               col_types = col_classes)
+
+message(paste0("Dataset has been read successfully with N = ", nrow(df), " rows"))
+
+# Format columns ---------------------------------------------------------------
```

Comment:
Is this necessary given you read the data in with `col_types` specified? If so, it should use the vectors of variables by type rather than re-doing the grep.

Reply:
True, should just use the vectors defined:
https://github.com/opensafely/post-covid-respiratory/blob/52a33bbaaefa71d58cc26648ea9a0f0ac6ea13f4/analysis/preprocess/preprocess_data.R#L65-L74

Reply:
Does it change any variables though as you read the data in with col_types specified? It might only be the dates that need updating here.

Reply:
Agree. Other than variables in the vector of date_cols, there is no need to update for `num_`,  `cat_` or `bin_`. 

I have a question here for `_birth_year`, taking `qa_num_birth_year` for example: 
1. We generated it in the variables_cohort.py: `qa_num_birth_year=patients.date_of_birth.year`
2. `qa_num_birth_year`  will be converted to numeric when identifying column classes: 
`num_cols <- c(grep("_num", all_cols, value = TRUE)`
4. These codes will then convert `qa_num_birth_year` to string. 
`          across(contains('_birth_year'), 
                 ~ format(as.Date(., origin = "1970-01-01"), "%Y")), `


Q: do we really use `qa_num_birth_year` as string? or actually numeric as implied from the name?

Reply:
`across(contains('_birth_year'),  ~ format(as.Date(., origin = "1970-01-01"), "%Y"))` makes it a date with just the year displayed. I think we probably want to keep this line.

Reply:
Back:
https://github.com/opensafely/post-covid-respiratory/blob/aca510e883769c341c549d86a033a7497b9c2eb9/analysis/preprocess/preprocess_data.R#L66-L70